File: dataset/ourMethod/OPENAI_gpt-5-2025-08-07/Task118.py
from typing import List, Tuple

MAX_INPUT_LENGTH = 1000

def is_length_safe(s: str) -> bool:
    return 0 < len(s) <= MAX_INPUT_LENGTH

def is_ascii(s: str) -> bool:
    try:
        s.encode("ascii")
        return True
    except UnicodeEncodeError:
        return False

def is_valid_ipv4(s: str) -> bool:
    if s is None or not is_length_safe(s) or not is_ascii(s) or (" " in s or "\t" in s):
        return False
    parts = s.split(".")
    if len(parts) != 4:
        return False
    for p in parts:
        if len(p) == 0 or len(p) > 3:
            return False
        if not p.isdigit():
            return False
        if len(p) > 1 and p[0] == "0":
            return False
        try:
            v = int(p)
        except ValueError:
            return False
        if v < 0 or v > 255:
            return False
    return True

def normalize_ipv4(s: str) -> str:
    if not is_valid_ipv4(s):
        return ""
    parts = s.split(".")
    return ".".join(str(int(p)) for p in parts)

def parse_ipv6_side(side: str) -> List[int] or None:
    res: List[int] = []
    if side == "":
        return res
    tokens = side.split(":")
    for t in tokens:
        if len(t) == 0 or len(t) > 4:
            return None
        try:
            v = int(t, 16)
        except ValueError:
            return None
        if v < 0 or v > 0xFFFF:
            return None
        res.append(v)
        if len(res) > 8:
            return None
    return res

def parse_ipv6_to_groups(s: str) -> List[int] or None:
    if s is None or not is_length_safe(s) or not is_ascii(s) or len(s) == 0:
        return None
    for c in s:
        if not (c == ":" or c.isdigit() or ("a" <= c.lower() <= "f")):
            return None
    dc_first = s.find("::")
    if dc_first != -1 and s.find("::", dc_first + 2) != -1:
        return None
    groups: List[int] = []
    if dc_first != -1:
        left = s[:dc_first]
        right = s[dc_first + 2:]
        left_groups = parse_ipv6_side(left)
        if left_groups is None:
            return None
        right_groups = parse_ipv6_side(right)
        if right_groups is None:
            return None
        total = len(left_groups) + len(right_groups)
        if total >= 8:
            return None
        groups.extend(left_groups)
        groups.extend([0] * (8 - total))
        groups.extend(right_groups)
    else:
        all_groups = parse_ipv6_side(s)
        if all_groups is None or len(all_groups) != 8:
            return None
        groups = all_groups
    if len(groups) != 8:
        return None
    return groups

def is_valid_ipv6(s: str) -> bool:
    return parse_ipv6_to_groups(s) is not None

def normalize_ipv6(s: str) -> str:
    groups = parse_ipv6_to_groups(s)
    if groups is None:
        return ""
    return ":".join(f"{g:04x}" for g in groups)

def detect_ip_type(s: str) -> str:
    if s is None or not is_length_safe(s) or not is_ascii(s):
        return "Invalid"
    if is_valid_ipv4(s):
        return "IPv4"
    if is_valid_ipv6(s):
        return "IPv6"
    return "Invalid"

def process_ip(s: str) -> str:
    t = detect_ip_type(s)
    if t == "IPv4":
        return f"Type: IPv4, Normalized: {normalize_ipv4(s)}"
    if t == "IPv6":
        return f"Type: IPv6, Normalized: {normalize_ipv6(s)}"
    return "Invalid"

def main() -> None:
    tests = [
        "192.168.1.1",
        "255.255.255.255",
        "01.2.3.4",
        "2001:0db8:85a3:0000:0000:8a2e:0370:7334",
        "2001:db8::8a2e:370:7334",
    ]
    for t in tests:
        print(f"{t} => {process_ip(t)}")

if __name__ == "__main__":
    main()

File: dataset/ourMethod/OPENAI_gpt-5-2025-08-07/Task118.java
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public final class Task118 {
    private static final int MAX_INPUT_LENGTH = 1000;

    public static String detectIPType(String input) {
        if (input == null) return "Invalid";
        if (!isLengthSafe(input)) return "Invalid";
        if (!isAscii(input)) return "Invalid";
        if (isValidIPv4(input)) return "IPv4";
        if (isValidIPv6(input)) return "IPv6";
        return "Invalid";
    }

    public static boolean isValidIPv4(String s) {
        if (!isLengthSafe(s) || s.indexOf(' ') >= 0 || s.indexOf('\t') >= 0) return false;
        String[] parts = s.split("\\.", -1);
        if (parts.length != 4) return false;
        for (String p : parts) {
            if (p.length() == 0 || p.length() > 3) return false;
            for (int i = 0; i < p.length(); i++) {
                char c = p.charAt(i);
                if (c < '0' || c > '9') return false;
            }
            if (p.length() > 1 && p.charAt(0) == '0') return false;
            try {
                int v = Integer.parseInt(p);
                if (v < 0 || v > 255) return false;
            } catch (NumberFormatException e) {
                return false;
            }
        }
        return true;
    }

    public static String normalizeIPv4(String s) {
        if (!isValidIPv4(s)) return "";
        String[] parts = s.split("\\.", -1);
        StringBuilder sb = new StringBuilder(15);
        for (int i = 0; i < 4; i++) {
            int v = Integer.parseInt(parts[i]);
            sb.append(v);
            if (i < 3) sb.append('.');
        }
        return sb.toString();
    }

    public static boolean isValidIPv6(String s) {
        return parseIPv6ToGroups(s) != null;
    }

    public static String normalizeIPv6(String s) {
        int[] groups = parseIPv6ToGroups(s);
        if (groups == null) return "";
        StringBuilder sb = new StringBuilder(39);
        for (int i = 0; i < 8; i++) {
            String hex = Integer.toHexString(groups[i]);
            for (int k = hex.length(); k < 4; k++) sb.append('0');
            sb.append(hex);
            if (i < 7) sb.append(':');
        }
        return sb.toString();
    }

    private static int[] parseIPv6ToGroups(String s) {
        if (s == null || !isLengthSafe(s) || !isAscii(s)) return null;
        if (s.isEmpty()) return null;
        // Only hex digits and ':' allowed
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            boolean ok = (c == ':') || (c >= '0' && c <= '9') ||
                         (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
            if (!ok) return null;
        }

        int dcFirst = s.indexOf("::");
        if (dcFirst != -1 && s.indexOf("::", dcFirst + 2) != -1) return null; // more than one '::'

        List<Integer> groups = new ArrayList<>(8);
        if (dcFirst >= 0) {
            String left = s.substring(0, dcFirst);
            String right = s.substring(dcFirst + 2);

            List<Integer> leftGroups = parseIPv6Side(left);
            if (leftGroups == null) return null;
            List<Integer> rightGroups = parseIPv6Side(right);
            if (rightGroups == null) return null;

            int total = leftGroups.size() + rightGroups.size();
            if (total >= 8) return null;
            groups.addAll(leftGroups);
            for (int i = 0; i < 8 - total; i++) groups.add(0);
            groups.addAll(rightGroups);
        } else {
            List<Integer> all = parseIPv6Side(s);
            if (all == null || all.size() != 8) return null;
            groups.addAll(all);
        }
        if (groups.size() != 8) return null;
        int[] out = new int[8];
        for (int i = 0; i < 8; i++) out[i] = groups.get(i);
        return out;
    }

    private static List<Integer> parseIPv6Side(String side) {
        List<Integer> res = new ArrayList<>(8);
        if (side.isEmpty()) return res; // zero groups
        String[] tokens = side.split(":", -1);
        for (String t : tokens) {
            if (t.isEmpty() || t.length() > 4) return null;
            int val = 0;
            for (int i = 0; i < t.length(); i++) {
                char c = t.charAt(i);
                int v;
                if (c >= '0' && c <= '9') v = c - '0';
                else if (c >= 'a' && c <= 'f') v = 10 + (c - 'a');
                else if (c >= 'A' && c <= 'F') v = 10 + (c - 'A');
                else return null;
                val = (val << 4) | v;
                if (val < 0 || val > 0xFFFF) return null;
            }
            res.add(val);
            if (res.size() > 8) return null;
        }
        return res;
    }

    private static boolean isLengthSafe(String s) {
        // Validate reasonable bounds
        return s.length() > 0 && s.length() <= MAX_INPUT_LENGTH;
    }

    private static boolean isAscii(String s) {
        byte[] bytes = s.getBytes(StandardCharsets.US_ASCII);
        return new String(bytes, StandardCharsets.US_ASCII).equals(s);
    }

    public static String processIP(String s) {
        String type = detectIPType(s);
        if ("IPv4".equals(type)) {
            return "Type: IPv4, Normalized: " + normalizeIPv4(s);
        } else if ("IPv6".equals(type)) {
            return "Type: IPv6, Normalized: " + normalizeIPv6(s);
        } else {
            return "Invalid";
        }
    }

    public static void main(String[] args) {
        String[] tests = new String[] {
            "192.168.1.1",
            "255.255.255.255",
            "01.2.3.4",
            "2001:0db8:85a3:0000:0000:8a2e:0370:7334",
            "2001:db8::8a2e:370:7334"
        };
        for (String t : tests) {
            System.out.println(t + " => " + processIP(t));
        }
    }
}